Bluetooth: host: Ignore failure to set passive scanner address

joerchan opened the pull request:
Ignore error when failing to set the passive scanner private address.
This can happen because in between the time we checked if the
advertiser was enabled and the time we execute the set random address
command the state of the advertiser could have changed in the
controller, and we only set the state once the command has completed
in the controller.

Fixes: #25672

Fix not setting NRPA before starting scanner.
Occurs with BT_PRIVACY=n and BT_EXT_ADV=y

joerchan committed:
Bluetooth: host: Ignore failure to set passive scanner address

Ignore error when failing to set the passive scanner private address.
This can happen because in between the time we checked if the
advertiser was enabled and the time we execute the set random address
command the state of the advertiser could have changed in the
controller, and we only set the state once the command has completed
in the controller.

Fixes: #25672

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -4113,7 +4113,14 @@ static int le_scan_set_random_addr(bool active_scan, uint8_t *own_addr_type)
+							     BT_ADV_ENABLED))) {
+					return err;
+				} else {
+					BT_WARN("Ignore failure to set address "
+						"for passive scan");
```
Maybe log the error value as well?

jhedberg replied:
And I'd use the form "Ignoring" rather than "Ignore"

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -4113,7 +4113,15 @@ static int le_scan_set_random_addr(bool active_scan, uint8_t *own_addr_type)
 			if (err) {
-				return err;
+				if (active_scan ||
+				    (adv && !atomic_test_bit(adv->flags,
```
Sorry, another thing I noticed: the top level condition for the branch that the `le_set_private_addr()` call is in is already checking for `adv != NULL` so checking for it again seems redundant here?

joerchan replied:
While processing this comment I realized that the above statement is wrong, there is an issue with the use of parenthesis.

I tested it with` west build -b native_posix tests/bluetooth/shell -- -DCONFIG_BT_PRIVACY=n -DCONFIG_BT_EXT_ADV=y`
Looking at the btmon there is no call te LE_Set_Random_Addr, before starting the scanner.
Added a fix, and now the NRPA is set.

I extracted it if-statement out to a helper function because repeating the same check was difficult at the indentation level.
The check should only be applied if the advertiser is using the random address, if the advertiser is using the LE_Set_Advertiser_Random_Address then this should not happen.

joerchan committed:
Bluetooth: host: Fix not setting NRPA before starting scanner

Fix not setting NRPA before starting scanner.
Occurs with BT_PRIVACY=n and BT_EXT_ADV=y

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -4083,6 +4083,21 @@ static void hci_cmd_status(struct net_buf *buf)
 }
 
 #if defined(CONFIG_BT_OBSERVER)
+static inline bool is_adv_using_rand_addr(void)
```
I'd drop the `inline` here and let the compiler take care of it (it will most certainly inline it anyway). IMO it's only in cases where you do careful profiling after the fact and discover you get a significant benefit out of explicit inlining that it's justified to do so.

joerchan replied:
Agree, removed it.
